ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/InsertsPublisher.java

```diff
@@ -0,0 +1,198 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.api.client;
+
+import java.util.ArrayDeque;
+import java.util.Objects;
+import java.util.Queue;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+/**
+ * A {@code org.reactivestreams.Publisher} suitable for use with the
+ * {@link Client#streamInserts(String, Publisher)} method. Rows for insertion are passed to the
+ * publisher via the {@link #accept(KsqlObject)} method, and buffered for delivery once the
+ * {@link Client#streamInserts} request is made and the server-side subscriber has been subscribed.
+ */
+public class InsertsPublisher implements Publisher<KsqlObject> {
+
+  /**
+   * The buffer max size indicator value used by the default constructor. See
+   * {@link #InsertsPublisher(int)} for how this value is used.
+   */
+  public static final int DEFAULT_BUFFER_MAX_SIZE = 200;
+
+  private Subscriber<? super KsqlObject> subscriber;
+  private final Queue<KsqlObject> buffer = new ArrayDeque<>();
+  private final int bufferMaxSize;
+  private long demand;
+  private Runnable drainHandler;
+  private volatile boolean cancelled;
+  private boolean complete;
+  private boolean shouldSendComplete;
+  private boolean sentComplete;
+
+  /**
+   * Creates an {@code InsertsPublisher}.
+   */
+  public InsertsPublisher() {
+    this(DEFAULT_BUFFER_MAX_SIZE);
+  }
+
+  /**
+   * Creates an {@code InsertsPublisher}.
+   *
+   * @param bufferMaxSize Indicative max number of elements to store in the buffer. Note that this
+   *                      value is not enforced, but it used to determine what to return from the
+   *                      {@link #accept(KsqlObject)} method so the caller can stop sending more
+   *                      rows and set a drainHandler to be notified when the buffer is cleared
+   */
+  public InsertsPublisher(final int bufferMaxSize) {
+    this.bufferMaxSize = bufferMaxSize;
+  }
+
+  /**
+   * Provides a new row for insertion. The publisher will attempt to deliver it to server endpoint,
+   * once the {@link Client#streamInserts} request has been made. The publisher will buffer the row
+   * internally if it can't deliver it immediately. Note that the row will be buffered even if the
+   * buffer is 'full', i.e., if number of elements is at least {@code bufferMaxSize}, as the
+   * {@code bufferMaxSize} value is not a hard limit. See {@link #InsertsPublisher(int)} for more.
+   *
+   * @param row the row to insert
+   * @return whether the internal buffer is 'full', i.e., if number of elements is at least
+   *         {@code bufferMaxSize}.
+   */
+  public synchronized boolean accept(final KsqlObject row) {
```

Comment:
Couldn't think of a way to make the methods in this class thread-safe without synchronizing all the public methods, but that doesn't feel right. Hopefully someone else has better ideas?